feat: version 3.0 - complete architecture rewrite

ProjectSky opened the pull request:
BREAKING CHANGE: This version is not backward compatible. All existing plugins must be rewritten.

Architecture:
- Restructure codebase with modular architecture (impl/, include/, natives/)
- Implement unified event system via RegisterEvent/UnregisterEvent
- Add DiscordResult for structured error handling and data access
- Implement unified callback pattern across all async operations
- Add template-based handle management with auto-cleanup
- Remove monolithic discord.cpp/queue.h in favor of modular components
- Split operations into dedicated modules (message, webhook, command, channel, role, etc.)

Features:
- Add Discord Components V2 support (sections, containers, media galleries, separators) (close #16)
- Add modal dialog support with text inputs
- Add poll support with multiple choice options
- Add scheduled events support
- Add DiscordHandleArray for batch operations
- Add comprehensive example scripts for all major features
- Add referenced message ID retrieval for reply messages (close  #17)

Build System:
- Migrate DPP and zlib to git submodules
- Add Windows static linking for DPP and OpenSSL (no manual DLL copying required) (resolve #3)
- Simplify Linux dependencies to build-essential and libssl-dev
- Add version.rc for Windows builds

Documentation:
- Update README with 3.0 changes and breaking change notice
- Add event reference documentation

ProjectSky committed:
feat: version 3.0 - complete architecture rewrite

BREAKING CHANGE: This version is not backward compatible. All existing plugins must be rewritten.

Architecture:
- Restructure codebase with modular architecture (impl/, include/, natives/)
- Implement unified event system via RegisterEvent/UnregisterEvent
- Add DiscordResult for structured error handling and data access
- Implement unified callback pattern across all async operations
- Add template-based handle management with auto-cleanup
- Remove monolithic discord.cpp/queue.h in favor of modular components
- Split operations into dedicated modules (message, webhook, command, channel, role, etc.)

Features:
- Add Discord Components V2 support (sections, containers, media galleries, separators) (completes #16)
- Add modal dialog support with text inputs
- Add poll support with multiple choice options
- Add scheduled events support
- Add DiscordHandleArray for batch operations
- Add comprehensive example scripts for all major features
- Add referenced message ID retrieval for reply messages (completes #17)

Build System:
- Migrate DPP and zlib to git submodules
- Add Windows static linking for DPP and OpenSSL (no manual DLL copying required) (resolve #3)
- Simplify Linux dependencies to build-essential and libssl-dev
- Add version.rc for Windows builds

Documentation:
- Update README with 3.0 changes and breaking change notice
- Add event reference documentation

ProjectSky committed:
fix build

Heapons commented:
@ProjectSky I wanted to get started on migrating my Discord plugin, but I've noticed that one cannot set avatar **_urls_** for webhooks. Reason why I ask is because I use the Steam Web API to retrieve player avatars.

Other than that; being able to fetch from local files is a neat feature, nonetheless.

ProjectSky commented:
> @ProjectSky I wanted to get started on migrating my Discord plugin, but I've noticed that one cannot set avatar **_urls_** for webhooks. Reason why I ask is because I use the Steam Web API to retrieve player avatars.

Missed this feature. Will add it back later.

chocochocotorta commented:
Aside from #20 issue, is it possible now to store a Handle from something like DiscordInteraction to use it for later use (for example, a EphemeralResponse on a Timer or a SQL thing)?
Not sure if there's an optional option to make the Handle not close automatically.

ProjectSky commented:
> Aside from #20 issue, is it possible now to store a Handle from something like DiscordInteraction to use it for later use (for example, a EphemeralResponse on a Timer or a SQL thing)? Not sure if there's an optional option to make the Handle not close automatically.

In the event, AutoFreeHandles can be used to control whether handles are automatically released, with the default value being true.

https://github.com/ProjectSky/sm-ext-discord/blob/bbdc704dc943effa0659b967004a90072609ffc5/scripting/include/discord/result.inc#L80-L87

chocochocotorta commented:
Oh, thank you so much!
I just readed it's explained at the discord_event_system.sp

Another thing, is it normal that the bot Starts and Resumes a lot?
<img width="321" height="738" alt="image" src="https://github.com/user-attachments/assets/0ef05443-b889-46c7-89e3-0e130ffae216" />

Not sure if it's because of high ping between the Game Server and the Discord API.

ProjectSky commented:
> Oh, thank you so much! I just readed it's explained at the discord_event_system.sp
> 
> Another thing, is it normal that the bot Starts and Resumes a lot? 
> Not sure if it's because of high ping between the Game Server and the Discord API.

Frequent triggering of resume events may be caused by poor network connectivity to the Discord gateway.

ProjectSky committed:
refactor: add array accessor methods and improve API consistency

Dependencies:
- Update DPP library (fixes poll vote events bug)
- Update zlib library

API Enhancements:
- Add array accessor methods returning DiscordHandleArray:
  * DiscordMessage.GetMentionedUserIds/RoleIds/ChannelIds()
  * DiscordEmoji.GetRoles()
  * DiscordGuildMember.GetRoles()
- Add DiscordMessage.GetMentionedChannelId(index) for individual channel access
- Add DiscordGuildMember.GetUserName/GetGlobalName() methods for cached user data
- Add DiscordGuildMember.HasPermissionEnum() helper method
- Add DiscordWebhook.SetAvatarFromUrl() method for URL-based avatar setting

API Changes:
- Remove guild_id parameter from DiscordUser constructor, FetchUser, and FindUser
- Remove DiscordUser.GetGuildId() method (guild context moved to DiscordGuildMember)
- Rename DiscordWebhook.SetAvatar to SetAvatarFromFile for clarity
- Add mutual exclusivity notes for SetAvatarFromFile and SetAvatarFromUrl

Memory Management:
- Add "Caller is responsible for closing the returned handle" notes to all methods returning DiscordHandleArray
- Add memory management notes to Find* methods (FindUser, FindChannel, FindRole, FindGuild)
- Document handle lifecycle for GetAttachment and GetReaction methods

Code Cleanup:
- Remove duplicate ImageType enum, use DiscordImageType consistently
- Update guild.CreateEmoji to use DiscordImageType instead of ImageType
- Standardize enum naming with Discord prefix (DiscordMessageFlags, DiscordStickerFormat, DiscordEventEntityType)
- Delete unused stage_operations files
- Change MSVC optimization flag from /Ox to /O2

Bug Fixes:
- Fix discord_webhook.sp to use "items" instead of "webhooks" result key
- Remove DPP poll vote bug warnings (fixed in updated DPP version)

Implementation:
- Add array accessor methods in entity headers (discord_message.h, discord_emoji.h, discord_guild_member.h)
- Add SetAvatarFromUrl implementation in discord_webhook.h
- Update all .inc files with consistent documentation
- Refactor user_natives.cpp to remove guild context handling

chocochocotorta commented:
I managed to try the new EditFromObject and it works very nice so far.
A thing i'm trying to understand, for some reason DiscordMessages created with "CreateWithContent" or a Embed Executed with a char containing a Role Mention or person mention will never mention both.
Not sure if in this case, the SetAllowedMentions method needs to be used or what, but Webhooks don't have it so that's why i'm confused.

ProjectSky commented:
> I managed to try the new EditFromObject and it works very nice so far. A thing i'm trying to understand, for some reason DiscordMessages created with "CreateWithContent" or a Embed Executed with a char containing a Role Mention or person mention will never mention both. Not sure if in this case, the SetAllowedMentions method needs to be used or what, but Webhooks don't have it so that's why i'm confused.

First, create a message object. Then, call SetAllowedMentions on it to configure mention permissions. Finally, send the message using ExecuteFromObject.

https://github.com/ProjectSky/sm-ext-discord/blob/d59fc92e82b245f50cf0775436909fd575f88400/scripting/include/discord/webhook.inc#L207-L235

chocochocotorta commented:
I don't know if this got broken with one of the updates (i forgot if i used the 3.0 with a 32 bits server) but, when trying to load a plugin (or loading the bot when using one of the examples with a bot token), the server sends this error using "sm-ext-discord-smlatest-ubuntu-22.04-2b9c184" (this version works fine on my other TF2 that's 64 bits, but not in the 32 bits one):

> L 05/05/2026 - 23:18:17: [SM] Exception reported: Could not add callback to forward
> L 05/05/2026 - 23:18:17: [SM] Blaming: discord_poll.smx
> L 05/05/2026 - 23:18:17: [SM] Call stack trace:
> L 05/05/2026 - 23:18:17: [SM]   [0] Discord.RegisterEvent
> L 05/05/2026 - 23:18:17: [SM]   [1] Line 43, d:\#Repositorios\sm-ext-discord-refactor-modular-architecture\scripting\discord_poll.sp::Command_Start

ProjectSky commented:
Due to an [update ](https://github.com/alliedmodders/sourcepawn/commit/91ab76ff197f31bf8f17841b5cccccf484d45c8a)in the SourcePawn API which added a new PushInt64 virtual function to ICallable, extensions need to be recompiled in order to work correctly

ProjectSky committed:
refactor: harden async callbacks and native API contracts

Add async callback tracking and main-thread re-resolution
Dispatch DPP events through the bounded task queue
Detach plugin callbacks during unload and client shutdown
Tighten SourcePawn native validation and int64 handling
Add timestamp helpers and Windows-safe numeric_limits usage
Update CI, build inputs, result enums, and vendored dependencies

ProjectSky committed:
fix: support SourceMod plugin context API changes
